### ISSUE TITLE
[dialog] Declarative indicator for dialog modality

### ISSUE THREAD

**Issue by lukewarlow** [openui/open-ui#920]
The invokers proposal is a brilliant addition and really helps solve some of the developer ergonomic issues with dialogs (along with much more).

However, it misses one key case as I mentioned in https://github.com/whatwg/html/issues/3567, you can't have a modal dialog load open with just HTML, you need JS to call showModal() or a user to click an invoker button.

My use case is that I sometimes have modal dialogs that are linked to a client side router where the URL controls their visibility. I currently call .showModal() on component mount. But if I could instead just do this declaratively that would remove the requirement for client JS execution (think SSR scenarios)

`<dialog open>` works for non-modals but there's no equivalent for modal dialogs.

**Comment by lukewarlow** [comment]
Happy to bikeshed etc but my proposal would be a new `modal` attribute. It wouldn't change anything by default it just acts as an indicator when used alongside the open attribute.

My thoughts on what `<dialog modal>` would do when called with JS.

`show()` - Throw an exception (I'm not fully set on this so could probably be persuaded either way).

`showModal()` - Work as expected.

This could also simplify the invoker proposal. Currently we have a [toggleModal](https://open-ui.org/components/invokers.explainer/#defaults) action, and Keith has added an additional `toggle` action inside of his Chromium CL.

If we added a new `modal` attribute, we could just have a single `toggle` action that would just work as expected?

Depending on the behaviour of `show()` JS function, we could even just have a `show` action and forget `showModal`?

**Comment by YummyBacon5** [comment]
> If we added a new modal attribute

Could `<dialog open="modal">` not work?

**Comment by lukewarlow** [comment]
> Could <dialog open="modal"> not work?

It would potentially for the open as modal on page load. Potentially as I'm not sure on changing boolean attribute to take values. I guess it was done with hidden so could be fine?

It wouldn't, however, additionally remove ambiguity around invoker actions though.

**Comment by keithamus** [comment]
> Potentially as I'm not sure on changing boolean attribute to take values. I guess it was done with hidden so could be fine?

Yes this is fine as it will become an enumerated attribute where the invalid/empty states map to <whatever the current behaviour is>. The problem is that the current behaviour will need a name.

**Comment by lukewarlow** [comment]
Naming things is hard, could that be another reason to use a new attribute for the modality instead. Then we get potential invoker benefits *and* we don't have to think of naming <whatever the current behaviour is>

**Comment by mfreed7** [comment]
Some quick references for naming and behavior, since a declarative "open when loaded" feature was initially part of Popover but was removed.

- https://github.com/openui/open-ui/issues/631
- https://github.com/whatwg/html/issues/7785#issuecomment-1294296444
- https://github.com/openui/open-ui/issues/500

**Comment by lukewarlow** [comment]
> Could <dialog open="modal"> not work?

Coming back to this since my original reason against this no longer holds (invoker commands aren't ambiguous anymore). I think this could be a neat way to do this for dialogs. `<dialog open>` would still be parsed as an open non-modal dialog. But `<dialog open="modal">` would be parsed as an open modal dialog.

We could maybe even make `dialog.showModal()` set it as `<dialog open="modal">` as far as I know that wouldn't break any code, especially if we make `dialog.open` always return a boolean value? `dialog[open]` would still match in CSS correctly for example.

We obviously still need the discussion on whether this is a good thing to have or not, but if we decide it is that seems like a fine API?

**Comment by ryantownsend** [comment]
@lukewarlow I could very much use this value existing.

I’m currently achieving it via just setting `open` in the HTML and then having JavaScript immediately call `close()` and `showModal()` during page load – it’s not ideal because you can see a flicker.

If it helps, my primary use-cases are:

1. A welcome modal triggered when a user first enters a section of an admin panel they’ve never visited previously, providing a dismissible explanation of the section.
2. More in-depth confirmation messages where a small toast-style notification won’t cut it, e.g. generating a new API key needs to present the one-time-visible key, provide context that other keys have been set to expire and link to documentation.

**Comment by mayank99** [comment]
This might sound like theoretical purity but… I think a separate `modal` attribute makes more sense than overloading the `open` attribute. Modality is a _property_ of the dialog, while openness is a _state_.

I don't expect the same dialog to be opened as non-modal _and_ modal during its lifecycle. I've always found it strange that it's not possible to determine at a glance whether a `<dialog>` will be modal when opened.

Another thing to consider: making the `open` attribute diverge from the `open` DOM property will cause a potential problem for JSX frameworks that work primarily with properties, i.e. `<dialog open="modal">` would typically translate to `dialog.open = "modal"` rather than `dialog.setAttribute("open", "modal")`. I'm sure frameworks can adapt, but it might take them a while if they deem this to be a breaking change for whatever reason.

**Comment by oliverjam** [comment]
This is actually quite annoying when building server-driven apps. Modern HTML/CSS is _so_ close to being able to create really nice interactive experiences without any client-side JS.

But I just hit this modal issue with form validation. If you submit a form from a dialog it's important to be able to redirect back with visible validation errors from the server. As far as I can tell there's no way to do this. I'd need to intercept the submission with client-side JS and render messages inline to avoid losing the open state on the page by redirecting.

**Comment by lukewarlow** [comment]
> I'd need to intercept the submission with client-side JS and render messages inline to avoid losing the open state on the page by redirecting.

Alternatively you'd have to have some mechanism to launch the dialog on page reload after a failed submission to the server.

**Comment by ryantownsend** [comment]
@lukewarlow I think @oliverjam's point is that if you have a form in a modal dialog, you can't have that hard-submit and SSR a response with the validation errors in the same `<dialog>` element and have it **declaratively** open on the next page load (hence needing/supporting the change described in this issue).

I have this same problem in one of my products. For now, I've 'hacked' around this with JavaScript as described above: my page will render the `<dialog>` with an `open` attribute, then JavaScript will `close()` and `showModal()` the element on page load to promote it to a modal dialog.

@oliverjam as an aside, you might want to follow @noamr's [Declarative Partial Updates proposal](https://github.com/WICG/declarative-partial-updates), I hope we'll get some form handling with that spec (I've written up this use-case in [this comment](https://github.com/WICG/declarative-partial-updates/issues/32#issuecomment-3696522405)). I wouldn't expect to see DPU support before this issue is resolved, but I think it suits your use-case nevertheless.
